### PR TITLE
Avail server updates

### PR DESCRIPTION
**Pull request by richATavail**
Added Heartbeats to the Avail server. This is an optional system that allows for keeping channel connections alive. It also adds a hook for performing close actions when closing a channel as well as capturing the origin of the close and the reason for the close.

This is the first phase of Avail server updates. I will be going through again to determine where close actions should be called.

### CODE REVIEW

File: avail-server/src/main/java/com/avail/server/io/TransportAdapter.kt [hunk 1]
```diff
@@ -87,4 +90,30 @@ interface TransportAdapter<T> : AutoCloseable
 	 *   A channel.
 	 */
 	fun sendClose(channel: AbstractTransportChannel<T>)
+
+	/**
+	 * A function that accepts a [DisconnectReason] and the underlying
```
Comment (toddATavail) [inline avail-server/src/main/java/com/avail/server/io/TransportAdapter.kt]: Revise this based on other changes made to `onChannelCloseAction` parameters and fields elsewhere, as mentioned above.

File: avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt [hunk 2]
```diff
@@ -86,14 +102,33 @@ import kotlin.experimental.xor
...
 	override val server: AvailServer,
 	internal val adapterAddress: InetSocketAddress,
-	internal val serverAuthority: String)
+	internal val serverAuthority: String,
+	private val heartbeatFailureThreshold: Int = 3,
+	private val heartbeatInterval: Long = 12000,
+	private val heartbeatTimeout: Long = 15000,
+	override val onChannelCloseAction:
+		(DisconnectReason, AbstractTransportChannel<AsynchronousSocketChannel>)
+			-> Unit = { _, _ -> /* Do nothing */})
```
Comment (toddATavail) [inline avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt]: Great use of default arguments. If it gets much more onerous, we might want to introduce a `WebSocketAdapterConfiguration`. I'll probably need to do that in order to introduce new features, but we can worry about that later.

File: avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt [hunk 3]
```diff
@@ -105,6 +140,16 @@ class WebSocketAdapter @Throws(IOException::class) constructor(
 		acceptConnections()
 	}
 
+	override val timer =
+		ScheduledThreadPoolExecutor(
+			1,
+			ThreadFactory { r ->
+				val thread = Thread(r)
+				thread.isDaemon = true
+				thread.name = "WebSocketAdapterTimer" + thread.id
```
Comment (toddATavail) [inline avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt]: Use a template string.

File: avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt [hunk 4]
```diff
@@ -1244,6 +1296,7 @@ class WebSocketAdapter @Throws(IOException::class) constructor(
 						}
 						else
 						{
+							channel.heartbeat.receiveHeartbeat()
 							sendPong(
 								strongChannel, bytes.toByteArray(), null, null)
```
Comment (toddATavail) [inline avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt]: Looks like we might want to give `sendPong` default arguments of `null` for the last two parameters.

File: avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt [hunk 5]
```diff
@@ -56,15 +58,44 @@ import java.nio.channels.AsynchronousSocketChannel
...
 internal class WebSocketChannel constructor(
 	override val adapter: WebSocketAdapter,
-	override val transport: AsynchronousSocketChannel)
+	override val transport: AsynchronousSocketChannel,
+	heartbeatFailureThreshold: Int,
+	heartbeatInterval: Long,
+	heartbeatTimeout: Long,
+	onChannelCloseAction:
+		(DisconnectReason, AbstractTransportChannel<AsynchronousSocketChannel>)
+			-> Unit)
```
Comment (toddATavail) [inline avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt]: We probably want to default these in the same way that you did for `WebSocketAdapter`.

File: avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt (hunk 2 above)
Comment (toddATavail) [reply avail-server/src/main/java/com/avail/server/io/WebSocketAdapter.kt]: One small request here — please introduce named `const val` fields in the companion to deanonymize the magic numbers `3`, `12000`, and `15000`. The empty arity-2 block is fine anonymous and in situ, however.

File: avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt (hunk 5 above)
Comment (toddATavail) [reply avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt]: You can use the same constants that you will introduce for the default arguments here.

File: avail-server/src/main/java/com/avail/server/io/SocketAdapter.kt [hunk 6]
```diff
@@ -69,14 +71,21 @@ import java.util.logging.Level
  *   An Avail server.
  * @param adapterAddress
  *   The socket address of the listener.
+ * @param onChannelCloseAction
+ *   A function that accepts a [DisconnectReason] and the underlying
+ *   [channel][AbstractTransportChannel] and answers `Unit` that is to be
+ *   called when the input channel is closed.
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-server/src/main/java/com/avail/server/io/SocketAdapter.kt]: I went with the following to more clearly specify that this is an injection point for implementation-specific logic for channel closing clean up:

```
   The custom action that is to be called when the input channel is closed in 
   order to support implementation-specific requirements for the closing of 
   a channel.
```

File: avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt (hunk 5 above)
Comment (richATavail) [reply avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt]: Actually, I specifically didn't do this as I believe we are treating the interface to the server for an AvailServer implementation is to be through the WebSocketAdapter, not WebSocketChannel. I don't believe a user using the server in their implementation would ever change the WebSocketAdapter code where a WebSocketChannel is created. For consistency and accessibility, I added these defaults at the WebSocketAdapter level as this is where you would override them with your own inputs. Additionally, with the addition of a configuration, you would also want to control these settings through configuration, not anywhere else.

**Comment by toddATavail** [comment]
Great answer. I agree.

**Comment by richATavail** [comment]
Going to try and open a new merge request

**Pull request closed without merging**